File: kmeans_assignment1/kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <limits.h>

#ifndef INFINITY 
#define INFINITY (__builtin_inf())
#endif

int k_means(char *filename, int k, int n, int d, int maxIter, double epsilon);

double** readFile(char *filePath, int d, int n);
double euclideanDistance(double *x, double *y, int d);
double* calMean(double **cluster, int num, int d);

int main(int argc, char *argv[]) {
    int k, n, d, maxIter=200;
    double epsilon=0.001;
    char filename[FILENAME_MAX];
    int result;
    if (argc >= 5) {
        sscanf(argv[1], "%d", &k);
        sscanf(argv[2], "%d", &n);
        sscanf(argv[3], "%d", &d);
        if (argc >=6) {
            sscanf(argv[4], "%d", &maxIter);
            sscanf(argv[5], "%s", filename);
        } else {
            maxIter = 200; 
            sscanf(argv[4], "%s", filename);
        }
    } 
    else {
        char maxIterInput[CHAR_MAX];
        printf("Enter filename, k, n, d: ");
        scanf("\"%99[^\"]\" %d %d %d", filename, &k, &n, &d);
        while (getchar() != '\n');
        printf("Enter max iteration: ");
        fgets(maxIterInput, sizeof(maxIterInput), stdin);
        if (maxIterInput[0] != '\n') {
            sscanf(maxIterInput,"%d", &maxIter);
        }
    }
    if ((int) k < 1 || (int) k>=n) {
        printf("Invalid number of clusters!\n");
        return 1;
    }
    if (n <= 1) {
        printf("Invalid number of points!\n");
        return 1;
    }
    if (d <= 0) {
        printf("Invalid dimension of point!\n");
        return 1;
    }
    if (maxIter <= 0 || maxIter >= 1000) {
        printf("Invalid maximum iteration!\n");
        return 1;
    }
    result =k_means(filename, k, n, d, maxIter, epsilon);
    return result;
}

int k_means(char *filename, int k, int n, int d, int maxIter, double epsilon){
    double **data=readFile(filename,d,n);
    double **centroids = (double **)malloc(k * sizeof(double *));
    int i,j,i1,b,z;
    int iter=0;
    if (data == NULL) {
        printf("An Error Has Occurred\n");
        return 1;
    }
    for (i = 0; i < k; i++) {
        centroids[i] = (double *)malloc(d* sizeof(double));
        for (j = 0; j < d; j++) {
            centroids[i][j] = data[i][j];
        }
    }
    while (iter < maxIter) {
        double **newCentroids = (double **)malloc(k * sizeof(double *));
        double ***clusters = (double ***)malloc(k * sizeof(double **));
        int *PointsInCluster = (int *)calloc(k, sizeof(int));
        int u,r,g,f,s,a,t;
        double e=0;
        for (u = 0; u < k; u++) {
            newCentroids[u] = (double *)malloc(d * sizeof(double));
        }
        for (r = 0; r < k; r++) {
            int l;
            clusters[r] = (double **)malloc(n * sizeof(double *));  
            for (l = 0; l < n; l++) {
                clusters[r][l]=(double *)malloc(d * sizeof(double));  
                }
            }
        for (g= 0; g< n; g++) {
            double minDist = INFINITY;
            int closestC;
            int w,ind;
            e=0;
            for (w = 0; w < k; w++) {
                double distance = euclideanDistance(data[g], centroids[w], d);
                if (distance < minDist) {
                    minDist = distance;
                    closestC = w;
                }
            }
            for(ind=0;ind<d;ind++){
                clusters[closestC][PointsInCluster[closestC]][ind] = data[g][ind];
            }
             PointsInCluster[closestC]++;
        }
        for (f = 0; f< k; f++) {
            double *mean=calMean(clusters[f],PointsInCluster[f], d);
            int m;
            for(m=0;m<d;m++){
                newCentroids[f][m]=mean[m];
        }
        }
        for (s = 0; s< k; s++) {
            e += euclideanDistance(newCentroids[s], centroids[s], d);
        }
        if (e < epsilon) {
            break;
        }
        for (a = 0; a < k; a++) {
            int v;
            for (v = 0; v < d; v++) {
                centroids[a][v] = newCentroids[a][v];
            }
        }
        for (t= 0; t< k; t++) {
            int h;
            free(newCentroids[t]);
            for (h = 0; h< PointsInCluster[t]; h++) {
                free(clusters[t][h]);
            }
            free(clusters[t]);
        }
        free(newCentroids);
        free(clusters);
        free(PointsInCluster);
        iter++;
    }
    for (i1 = 0; i1< k; i1++) {
        int j1;
        for (j1 = 0; j1< d; j1++) {
            printf("%.4f ", centroids[i1][j1]);
        }
        printf("\n");
    }
    for (b= 0; b < n; b++) {
        free(data[b]);
    }
    free(data);
    for (z = 0; z < k; z++) {
        free(centroids[z]);
    }
    free(centroids);
    return 0;
}

double** readFile(char *filePath, int d, int n) {
    double **data = (double **)malloc(n * sizeof(double *));
    int q;
    FILE *file = fopen(filePath, "r");
    if (file == NULL) {
        free(data);
        return NULL;
    }
    for(q = 0; q < n; q++){
        double *row = (double *)malloc(d * sizeof(double));
        char newline;
        int p,s,z;
        if (row == NULL) {
            fclose(file);
            for (p = 0; p < q; p++) {
                free(data[p]);
            }
            free(data);
            return NULL; 
        }
        for (s = 0; s < d; s++) {
            int t;
            if (fscanf(file, "%lf,", &row[s]) != 1) {
                fclose(file);
                for (t= 0; t<= q; t++) {
                    free(data[t]);
                 }
                free(data);
                free(row);
                return NULL; 
            }
        }
        if (fscanf(file, "%c", &newline) != 1 || newline != '\n') {
            fclose(file);
            for (z = 0; z <= q; z++) {
                free(data[z]);
            }
            free(data);
            free(row);
            return NULL; 
        }

        data[q] = row;
    }

    fclose(file);
    return data;
} 
double euclideanDistance(double *x, double *y, int d) {
    double sum = 0.0;
    int i;
    for (i = 0; i < d; i++) {
        sum =sum +((x[i] - y[i]) * (x[i] - y[i]));
    }
    return sqrt(sum);
}

double* calMean(double **cluster, int num, int d) {
    double *mean = (double *)malloc(d * sizeof(double));
    int i,p,y;
    for (i = 0; i < d; i++) {
        mean[i] = 0.0;
    }
    for (p = 0; p < num; p++) {
        int j;
        for (j = 0; j < d; j++) {
            mean[j] += cluster[p][j];
        }
    }
    for(y = 0;y< d;y++) {
        mean[y] /= num;
    }
    return mean;
}

File: kmeans_assignment1/kmeans.py
import numpy as np
import sys
def k_means(filePath,k,n,d,iterNum=200,epsilon=0.001):
    try :
        data=readFile(filePath,d,n);
        iter=0;
        if(k<1 or k>n or not isinstance(k, int)):
            print("Invalid number of clusters!");
            return 1;
        if (n<=1 or not isinstance(n, int)):
            print("Invalid number of points!");
            return 1;
        if (d<=0 or not isinstance(d, int)):
            print("Invalid dimension of point!");
            return 1;
        if(iterNum<1 or iterNum>1000 or not isinstance(iterNum, int)):
            print("Invalid maximum iteration!");
            return 1;
        centroids=data[:k];
        while iter<iterNum:
            newCentroids=[];
            clusters =[[] for i in range(k)];
            for point in data:
                minDist=np.inf;
                for i in range(k):
                    distance=euclideanDistance(point,centroids[i]);
                    if (distance<minDist):
                        minDist=distance;
                        closeC=i;
                clusters[closeC].append(point);
            for cluster in clusters:
                newCentroids.append(np.mean(cluster,axis=0));
            e=0;
            for c in range(k):
                e+=euclideanDistance(newCentroids[c],centroids[c]);
            if(e<epsilon):
                break;
            centroids=newCentroids;
            iter+=1;
            
        formattedCentroids = [[f'%.4f' % float(val) for val in sublist] for sublist in centroids];
        returnValue = [','.join(sublist) for sublist in formattedCentroids];
        for value in returnValue :
            print(value,'\n');
        return 0;
    except Exception as e:
        print("An Error Has Occurred");
        return 1;


def euclideanDistance(x,y): 
    result=np.sqrt(np.sum((x-y)**2));
    return result;

def readFile(filePath, d, n):
    Fdata = []
    try:
        with open(filePath, 'r') as file:
            for line in file:
                values = line.strip().split(",")
                point = []
                for value in values:
                    point.append(float(value))
                if len(point) != d:
                    print("Error: Dimension of point doesn't match specified value (d).")
                    return None
                Fdata.append(point)
            if len(Fdata) < n:
                print("Error: Number of points in file is less than specified value (n).")
                return None
            if len(Fdata) > n:
                print("Warning: Number of points in file is greater than specified value (n).")
                Fdata = Fdata[:n]
        return np.array(Fdata)
    except FileNotFoundError:
        print(f"Error: File '{filePath}' not found.")
        return None;
    except Exception as e:
        print(f"An error occurred while reading the file: {e}")
        return None;

def main():
    if len(sys.argv) >= 5:
        k = int(sys.argv[1])
        n = int(sys.argv[2])
        d = int(sys.argv[3])
        if len(sys.argv) >= 6:
            filePath = sys.argv[5]
            iterNum = int(sys.argv[4])
        else:
            filePath = sys.argv[4]
            iterNum=200
    else:
        input_values = input("Enter k n d maximum iteration and file path separated by spaces: ").split();
        k, n, d = map(int, input_values[:3]);
        iterNum = int(input_values[3]) if len(input_values) > 4 else 200;
        filePath=input_values[-1];
    result=k_means(filePath,k,n,d,iterNum);

if __name__ == "__main__":
    main()  